File: components/credentials/src/main/java/org/cloudfoundry/credhub/domain/CertificateCredentialVersion.java
package org.cloudfoundry.credhub.domain;

import java.time.Instant;
import java.util.Objects;

import org.apache.commons.lang3.StringUtils;
import org.cloudfoundry.credhub.credential.CertificateCredentialValue;
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData;
import org.cloudfoundry.credhub.requests.GenerationParameters;
import org.cloudfoundry.credhub.utils.CertificateReader;

public class CertificateCredentialVersion extends CredentialVersion {

  private final CertificateCredentialVersionData delegate;
  private CertificateReader parsedCertificate;

  public CertificateCredentialVersion(final CertificateCredentialVersionData delegate) {
    super(delegate);
    this.delegate = delegate;
    this.setCertificate(delegate.getCertificate());
  }

  public CertificateCredentialVersion(final String name) {
    this(new CertificateCredentialVersionData(name));
  }

  public CertificateCredentialVersion() {
    this(new CertificateCredentialVersionData());
  }

  public CertificateCredentialVersion(final CertificateCredentialValue certificate, final Encryptor encryptor) {
    this();
    this.setEncryptor(encryptor);
    this.setCa(certificate.getCa());
    this.setPrivateKey(certificate.getPrivateKey());
    this.setCaName(certificate.getCaName());
    this.setCertificate(certificate.getCertificate());
    this.setTransitional(certificate.isTransitional());
    this.setExpiryDate(certificate.getExpiryDate());
    this.setCertificateAuthority(certificate.isCertificateAuthority());
    this.setTrustedCa(certificate.getTrustedCa());
    this.setSelfSigned(certificate.isSelfSigned());
    this.setGenerated(certificate.getGenerated());
  }

  public CertificateReader getParsedCertificate() {
    return this.parsedCertificate;
  }

  public String getCa() {
    return delegate.getCa();
  }

  public void setCa(final String ca) {
    delegate.setCa(ca);
  }

  public String getCertificate() {
    return delegate.getCertificate();
  }

  public void setCertificate(final String certificate) {
    delegate.setCertificate(certificate);
    if (StringUtils.isNotEmpty(delegate.getCertificate())) {
      parsedCertificate = new CertificateReader(certificate);
    }
  }

  public String getTrustedCa() {
    return delegate.getTrustedCa();
  }

  public void setTrustedCa(final String trustedCa) {
    delegate.setTrustedCa(trustedCa);
  }

  public String getPrivateKey() {
    return (String) super.getValue();
  }

  public void setPrivateKey(final String privateKey) {
    if (privateKey != null) {
      super.setValue(privateKey);
    }
  }

  public String getCaName() {
    return delegate.getCaName();
  }

  public void setCaName(final String caName) {
    delegate.setCaName(caName);
  }

  @Override
  public String getCredentialType() {
    return delegate.getCredentialType();
  }

  @Override
  public void rotate() {
    final String decryptedPrivateKey = this.getPrivateKey();
    this.setPrivateKey(decryptedPrivateKey);
  }

  @Override
  public boolean matchesGenerationParameters(final GenerationParameters generationParameters) {
    if (generationParameters == null) {
      return true;
    }

    final CertificateGenerationParameters parameters = (CertificateGenerationParameters) generationParameters;
    final CertificateGenerationParameters existingGenerationParameters = new CertificateGenerationParameters(getParsedCertificate(), getCaName());
    return existingGenerationParameters.equals(parameters);
  }

  public void setTransitional(final boolean transitional) {
    delegate.setTransitional(transitional);
  }

  public Instant getExpiryDate() {
    return delegate.getExpiryDate();
  }

  public void setExpiryDate(final Instant expiryDate) {
    delegate.setExpiryDate(expiryDate);
  }

  public boolean isVersionTransitional() {
    return delegate.isTransitional();
  }

  public boolean isSelfSigned() {
    return delegate.isSelfSigned();
  }

  public void setSelfSigned(final boolean isSelfSigned) {
    delegate.setSelfSigned(isSelfSigned);
  }

  public boolean isCertificateAuthority() {
    return delegate.isCertificateAuthority();
  }

  public void setCertificateAuthority(final boolean certificateAuthority) {
    delegate.setCertificateAuthority(certificateAuthority);
  }

  public Boolean getGenerated() {
    return delegate.getGenerated();
  }

  public void setGenerated(final Boolean generated) {
    delegate.setGenerated(generated);
  }

  @Override
  public GenerationParameters getGenerationParameters() {
    return null;
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    final CertificateCredentialVersion that = (CertificateCredentialVersion) o;
    return Objects.equals(delegate, that.delegate)
      && Objects.equals(getName(), that.getName())
      && Objects.equals(getUuid(), that.getUuid())
      && Objects.equals(getVersionCreatedAt(), that.getVersionCreatedAt());
  }

  @Override
  public int hashCode() {
    return Objects.hash(delegate);
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/domain/CertificateCredentialVersion.kt
package org.cloudfoundry.credhub.domain

import org.apache.commons.lang3.StringUtils
import org.cloudfoundry.credhub.credential.CertificateCredentialValue
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData
import org.cloudfoundry.credhub.requests.GenerationParameters
import org.cloudfoundry.credhub.utils.CertificateReader
import java.time.Instant
import java.util.Objects

class CertificateCredentialVersion(delegate: CertificateCredentialVersionData) : CredentialVersion(delegate) {
    lateinit var parsedCertificate: CertificateReader
        private set

    var ca: String?
        get() = (delegate as CertificateCredentialVersionData).ca
        set(ca) {
            (delegate as CertificateCredentialVersionData).ca = ca
        }

    var certificate: String?
        get() = (delegate as CertificateCredentialVersionData).certificate
        set(certificate) {
            (delegate as CertificateCredentialVersionData).certificate = certificate
            if (StringUtils.isNotEmpty((delegate as CertificateCredentialVersionData).certificate)) {
                parsedCertificate = CertificateReader(certificate)
            }
        }

    var trustedCa: String?
        get() = (delegate as CertificateCredentialVersionData).trustedCa
        set(trustedCa) {
            (delegate as CertificateCredentialVersionData).trustedCa = trustedCa
        }

    var privateKey: String?
        get() = super.getValue() as String?
        set(privateKey) {
            if (privateKey != null) {
                super.setValue(privateKey)
            }
        }

    var caName: String?
        get() = (delegate as CertificateCredentialVersionData).caName
        set(caName) {
            (delegate as CertificateCredentialVersionData).caName = caName
        }

    var expiryDate: Instant?
        get() = (delegate as CertificateCredentialVersionData).expiryDate
        set(expiryDate) {
            (delegate as CertificateCredentialVersionData).expiryDate = expiryDate
        }

    val isVersionTransitional: Boolean
        get() = (delegate as CertificateCredentialVersionData).isTransitional()

    var isSelfSigned: Boolean
        get() = (delegate as CertificateCredentialVersionData).isSelfSigned
        set(isSelfSigned) {
            (delegate as CertificateCredentialVersionData).isSelfSigned = isSelfSigned
        }

    var isCertificateAuthority: Boolean
        get() = (delegate as CertificateCredentialVersionData).isCertificateAuthority
        set(certificateAuthority) {
            (delegate as CertificateCredentialVersionData).isCertificateAuthority = certificateAuthority
        }

    var generated: Boolean?
        get() = (delegate as CertificateCredentialVersionData).generated
        set(generated) {
            (delegate as CertificateCredentialVersionData).generated = generated
        }

    init {
        this.certificate = delegate.certificate
    }

    constructor(name: String) : this(CertificateCredentialVersionData(name)) {}

    constructor(certificate: CertificateCredentialValue, credentialName: String, encryptor: Encryptor) : this(credentialName) {
        this.setEncryptor(encryptor)
        this.ca = certificate.ca
        this.privateKey = certificate.privateKey
        this.caName = certificate.caName
        this.certificate = certificate.certificate
        this.setTransitional(certificate.transitional)
        this.expiryDate = certificate.expiryDate
        this.isCertificateAuthority = certificate.certificateAuthority
        this.trustedCa = certificate.trustedCa
        this.isSelfSigned = certificate.selfSigned
        this.generated = certificate.generated
    }

    override fun getCredentialType(): String {
        return delegate.credentialType
    }

    override fun rotate() {
        val decryptedPrivateKey = this.privateKey
        this.privateKey = decryptedPrivateKey
    }

    override fun matchesGenerationParameters(generationParameters: GenerationParameters?): Boolean {
        if (generationParameters == null) {
            return true
        }

        val parameters = generationParameters as CertificateGenerationParameters?
        val existingGenerationParameters = CertificateGenerationParameters(parsedCertificate, caName)
        return existingGenerationParameters == parameters
    }

    fun setTransitional(transitional: Boolean) {
        (delegate as CertificateCredentialVersionData).transitional = transitional
    }

    override fun getGenerationParameters(): GenerationParameters? {
        return null
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as CertificateCredentialVersion?
        return (
            delegate == that!!.delegate &&
                name == that.name &&
                uuid == that.uuid &&
                versionCreatedAt == that.versionCreatedAt
            )
    }

    override fun hashCode(): Int {
        return Objects.hash(delegate)
    }
}
